src/Avalonia.Controls/Border.cs

```diff
@@ -120,29 +110,205 @@ protected override Size ArrangeOverride(Size finalSize)
         {
             if (Child != null)
             {
-                var padding = Padding + new Thickness(BorderThickness);
+                var padding = Padding + BorderThickness;
                 Child.Arrange(new Rect(finalSize).Deflate(padding));
             }
 
+            _borderRenderer.Update(finalSize, BorderThickness, CornerRadius);
+
             return finalSize;
         }
 
         internal static Size MeasureOverrideImpl(
             Size availableSize,
             IControl child,
             Thickness padding,
-            double borderThickness)
+            Thickness borderThickness)
         {
-            padding += new Thickness(borderThickness);
+            padding += borderThickness;
 
             if (child != null)
             {
                 child.Measure(availableSize.Deflate(padding));
                 return child.DesiredSize.Inflate(padding);
             }
-            else
+
+            return new Size(padding.Left + padding.Right, padding.Bottom + padding.Top);
+        }
+
+        internal class BorderRenderer
+        {
+            private bool _useComplexRendering;
+            private StreamGeometry _backgroundGeometryCache;
+            private StreamGeometry _borderGeometryCache;
+
+            public void Update(Size finalSize, Thickness borderThickness, CornerRadius cornerRadius)
             {
-                return new Size(padding.Left + padding.Right, padding.Bottom + padding.Top);
+                if (borderThickness.IsUniform && cornerRadius.IsUniform)
+                {
+                    _backgroundGeometryCache = null;
+                    _borderGeometryCache = null;
+                    _useComplexRendering = false;
+                }
+                else
+                {
+                    _useComplexRendering = true;
+
+                    var boundRect = new Rect(finalSize);
+                    var innerRect = new Rect(borderThickness.Left, borderThickness.Top,
+                        boundRect.Width - borderThickness.Right, boundRect.Height - borderThickness.Bottom);
+
+                    StreamGeometry backgroundGeometry = null;
+
+                    if (!boundRect.Width.Equals(0) && !boundRect.Height.Equals(0))
+                    {
+                        backgroundGeometry = new StreamGeometry();
+
+                        if (cornerRadius.IsEmpty)
```

Comment:
Why there are two identical if-else branches?

Reply:
Good catch. Those checks are not needed anymore.